src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const StateCopy = { ...state };
+  const EmptyArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(StateCopy, action.extraData);
+        EmptyArray.push(Object.assign({}, StateCopy));
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete StateCopy[key];
+        }
+        EmptyArray.push(Object.assign({}, StateCopy));
+        break;
+
+      case 'clear':
+        for (const StateKey in StateCopy) {
+          if (StateCopy.hasOwnProperty(StateKey)) {
+            delete StateCopy[StateKey];
+          };
+        }
+        EmptyArray.push(Object.assign({}, StateCopy));
+        break;
+
+      default:
+        return 'Whoops, something bad happened';
```

Comment:
In case of an invalid action type, you shouldn't return a string. It's better to throw an error or ignore the invalid action.

`throw new Error('Whoops, something bad happened')`